Support secrets for Route53 Access Key IDs

Compy opened the pull request:
Signed-off-by: Compy <hello@86pixels.com>

<!--

Thanks for opening a pull request! Here are some tips to get everything merged smoothly:

1. Read our contributor guidelines: https://cert-manager.io/docs/contributing/

2. Make sure your commits are signed off: https://cert-manager.io/docs/contributing/sign-off/

3. If the PR is unfinished, raise it as a draft or prefix the title with "WIP:" so it's clear to everyone.

4. Be sure to allow edits from maintainers so it's easier for us to help: https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

-->

### Pull Request Motivation

Implements the feature requested in #3568 
Currently if you want to use DNS01 validation with Route53, you specify the AWS Access Key ID in plaintext by using the `dns01.route53.accessKeyID` field. This introduces `accessKeyIDSecretRef` and is syntactically identical to `secretAccessKeySecretRef`

A new example:
```yaml
      dns01:
        route53:
          region: us-east-1
          secretAccessKeySecretRef:
            name: aws-route53-credentials
            key: secretAccessKey
          accessKeyIDSecretRef:
            name: aws-route53-credentials
            key: secretAccessKeyID
```

Specifying the access key ID using the current `accessKeyID` field is still supported.

### Kind
feature

### Release Note

```release-note
Added support for pulling both AWS access key IDs and secret keys from Kubernetes secrets
```

jsoref commented on deploy/crds/crd-challenges.yaml:
```diff
@@ -329,6 +329,18 @@ spec:
                             accessKeyID:
                               description: 'The AccessKeyID is used for authentication. If not set we fall-back to using env vars, shared credentials file or AWS Instance metadata see: https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials'
                               type: string
+                            accessKeyIDSecretRef:
+                              description: The SecretAccessKeyID is used for authentication. If not set we fall-back to using env vars, shared credentials file or AWS Instance metadata https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials
+                              type: object
+                              required:
+                                - name
+                              properties:
+                                key:
+                                  description: The key of the entry in the Secret resource's `data` field to be used. Some instances of this field may be defaulted, in others it may be required.
+                                  type: string
```
So very curious about the inconsistent use of quotation marks.

Compy replied:
This was copied from the `secretAccessKeySecretRef` attributes. The quotations all throughout the file seem inconsistent. I think its a good idea! I just think we need to resolve it in another issue entirely so we don't cloud the core feature modifications that this PR presents.

jsoref commented on deploy/crds/crd-challenges.yaml:
```diff
@@ -329,6 +329,18 @@ spec:
+                                - name
+                              properties:
+                                key:
+                                  description: The key of the entry in the Secret resource's `data` field to be used. Some instances of this field may be defaulted, in others it may be required.
```
Not a huge fan of this text.

Could someone explain more? Is there a url that explains more?

jsoref commented on deploy/crds/crd-clusterissuers.yaml:
```diff
@@ -364,6 +364,18 @@ spec:
                                   accessKeyID:
                                     description: 'The AccessKeyID is used for authentication. If not set we fall-back to using env vars, shared credentials file or AWS Instance metadata see: https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials'
                                     type: string
+                                  accessKeyIDSecretRef:
+                                    description: The SecretAccessKeyID is used for authentication. If not set we fall-back to using env vars, shared credentials file or AWS Instance metadata https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials
```
The first thing used `see:` this doesn't.

Not saying I like either pattern, but the dissonance...

Compy commented:
@jakexks  Thanks for the initial review!  In response to your suggestions I have:

- Added a validation that enforces specifying either accessKeyID or accessKeyIDSecretRef, but not both. https://github.com/cert-manager/cert-manager/blob/5ae36cc39628948a51aff856a3a68b161be673f4/pkg/issuer/acme/dns/dns.go#L298-L300
- Added a unit test that tests route53 configuration with accessKeyIDSecretRef  https://github.com/cert-manager/cert-manager/blob/5ae36cc39628948a51aff856a3a68b161be673f4/pkg/issuer/acme/dns/dns_test.go#L407-L463

Compy commented:
/retest

munnerz commented on pkg/issuer/acme/dns/dns.go:
```diff
@@ -292,6 +292,30 @@ func (s *Solver) solverForChallenge(ctx context.Context, issuer v1.GenericIssuer
+
+		// For route53, you must specify either an AccessKeyID or a secret reference to an AccessKeyID
+		// but not both.
+		if providerConfig.Route53.AccessKeyID != "" && providerConfig.Route53.SecretAccessKeyID.Name != "" {
```
Whilst we should also have this logic here, it _also_ needs to exist in the validation webhook/as part of our API - this will mean that if a user attempts to submit a resource that _does_ set both, it won't be admitted into the apiserver in the first place.

These changes should be made in the API validation package here: https://github.com/cert-manager/cert-manager/blob/c16b3cca7b418ba0d0b2bf1066514b8762984517/internal/apis/certmanager/validation/issuer.go#L401-L407

and you can add some test cases for this code here: https://github.com/cert-manager/cert-manager/blob/c16b3cca7b418ba0d0b2bf1066514b8762984517/internal/apis/certmanager/validation/issuer_test.go#L708-L715

munnerz commented on pkg/issuer/acme/dns/dns.go:
```diff
@@ -292,6 +292,30 @@ func (s *Solver) solverForChallenge(ctx context.Context, issuer v1.GenericIssuer
+		// Default to the AccessKeyID literal in the configuration
+		secretAccessKeyID := strings.TrimSpace(providerConfig.Route53.AccessKeyID)
+
+		// If the AccessKeyID secret reference option is defined, override the secretAccessKeyID variable
```
We shouldn't ever allow having both specified, so perhaps this code/text can be updated to reflect that :) (I think returning an error is valid given this isn't really an expected state to be in once the validation is done in the webhook/validation code)

munnerz commented on pkg/issuer/acme/dns/dns.go:
```diff
@@ -292,6 +292,30 @@ func (s *Solver) solverForChallenge(ctx context.Context, issuer v1.GenericIssuer
+				return nil, nil, fmt.Errorf("error getting route53 secret access key id: %s", err)
+			}
+
+			secretAccessKeyIDBytes, ok := secretAccessKeyIDSecret.Data[providerConfig.Route53.SecretAccessKeyID.Key]
```
If `SecretAccessKeyID.Key` is not set (i.e. the user didn't set it in the API request), we should fail a lot earlier than this.

Can we add additional validation to ensure this `.Key` field _is_ set if the `.Name` field is specified (and vice-versa?).

This could either be done by defining a new struct type other than `SecretKeySelector` in the APIs package, and using a `// +kubebuilder:default` marker on the `Key` field, or otherwise using a defaulting function in this file: https://github.com/cert-manager/cert-manager/blob/c16b3cca7b418ba0d0b2bf1066514b8762984517/internal/apis/certmanager/v1/defaults.go

munnerz replied:
We should also check here prior to using `.Key` that it *is* non-empty, and fail with a more specific/clear error in that case :)

munnerz reviewed (commented):
We should shift as much of the validation for this field (and how it relates with the existing `accessKeyID` field) into the webhook/CRDs validation functions/schema so that we don't have resources in invalid states/configurations :)

irbekrm commented:
/milestone v1.9

jakexks commented:
/kind feature

jakexks commented:
/test pull-cert-manager-make-test
/test pull-cert-manager-upgrade

jakexks reviewed (approved):
/lgtm
